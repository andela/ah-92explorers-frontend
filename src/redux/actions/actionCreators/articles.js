/* istanbul ignore file */
/* eslint-disable consistent-return */
import '@babel/polyfill';
import dotenv from 'dotenv';
import axios from 'axios';
import jwtDecode from 'jwt-decode';
import {
  CREATE_ARTICLE,
  GET_ARTICLE,
  FAILED_ARTICLE_CREATION,
  FAILED_ARTICLE_UPDATE,
  UPDATE_ARTICLE,
  SET_LOADING,
  GET_FEED,
  ARTICLE_GET_FAIL,
  GET_RATING,
} from '../actionTypes';
import fetchImage from '../../../helpers/createDisplayImage';
import terrestial from '../../../assets/icons/terrestial.jpg';

dotenv.config();

const success = (article, owner, authenticated) => ({
  type: GET_ARTICLE,
  payload: {
    article,
    owner,
    authenticated,
  },
});

const performAction = (payload) => ({
  type: GET_RATING,
  payload,
});

export const setLoading = (data) => (dispatch) => {
  dispatch({
    type: SET_LOADING,
    payload: data,
  });
};

export const publishArticle = (articleData) => async (dispatch) => {
  try {
    const {
      title,
      body,
      tagList,
    } = articleData;
    const tags = tagList[0];
    if (!title || title.length < 5 || !body) {
      return dispatch({
        type: FAILED_ARTICLE_CREATION,
        payload: {
          error: 'Did you forget to write a title/body!!!',
        },
      });
    }
    const newData = {
      title,
      body,
      tagList: tags === 'Tag' ? null : tags,
    };
    const article = await axios.post(`${process.env.APP_URL_BACKEND}/api/articles`, newData, {
      headers: {
        Authorization: localStorage.getItem('jwtToken'),
      },
    });
    dispatch({
      type: CREATE_ARTICLE,
      payload: article.data,
    });
  } catch (error) {
    dispatch({
      type: FAILED_ARTICLE_CREATION,
      payload: error.response.data,
    });
  }
};

export const updateArticle = (articleData, slug) => async (dispatch) => {
  try {
    const {
      title,
      body,
      tagList,
    } = articleData;
    const tags = tagList[0];
    if (!title || title.length < 5 || !body) {
      return dispatch({
        type: FAILED_ARTICLE_UPDATE,
        payload: {
          error: 'Did you forget to write a title/body!!!',
        },
      });
    }
    const newData = {
      title,
      body,
      tagList: tags === 'Tag' ? null : tags,
    };
    const article = await axios.put(`${process.env.APP_URL_BACKEND}/api/articles/${slug}`, newData, {
      headers: {
        Authorization: localStorage.getItem('jwtToken'),
      },
    });
    dispatch({
      type: UPDATE_ARTICLE,
      payload: article.data,
    });
  } catch (error) {
    dispatch({
      type: FAILED_ARTICLE_UPDATE,
      payload: error.response.data,
    });
  }
};

export const deleteArticle = (slug) => async (dispatch) => {
  try {
    const del = await axios.delete(`${process.env.APP_URL_BACKEND}/api/articles/${slug}/`, {
      headers: {
        Authorization: localStorage.getItem('jwtToken'),
      },
    });
    if (del.status === 204) {
      window.location.replace('/');
    }
  } catch (error) {
    return dispatch({
      type: ARTICLE_GET_FAIL,
      error: 'something went wrong',
    });
  }
};

export const getArticle = (slug) => async (dispatch) => {
  const token = localStorage.getItem('jwtToken') && jwtDecode(localStorage.getItem('jwtToken'));
  let authenticated = false;
  try {
    if (token && token.email) authenticated = true;
    const articles = await axios.get(`${process.env.APP_URL_BACKEND}/api/articles/${slug}/`);
    if (articles.status === 404) {
      window.location.replace('/not-found');
    }
    const owner = token && token.id === articles.data.article.author.id;
    return dispatch(success(articles.data.article, owner, authenticated));
  } catch (error) {
    if (error.response && error.response.status === 404) window.location.replace('/not-found');
    return dispatch({
      type: ARTICLE_GET_FAIL,
      error: 'something went wrong',
    });
  }
};

export const getRating = (slug) => async (dispatch) => {
  const token = localStorage.getItem('jwtToken') && jwtDecode(localStorage.getItem('jwtToken'));
  let authenticated = false;
  try {
    if (token && token.email) authenticated = true;
    const res = await axios.get(`${process.env.APP_URL_BACKEND}/api/article/${slug}/rating`);
    dispatch(performAction(res.data, authenticated));
  } catch (error) {
    return dispatch(performAction('something went wrong'));
  }
};

export const getFeed = (page) => async (dispatch) => {
  dispatch(setLoading(true));
  const article = await axios.get(`${process.env.APP_URL_BACKEND}/api/articles/feed?page=${!page || page === undefined ? 1 : page}&limit=13`);
  const {
    articles,
  } = article.data;
  const newArticles = [];

  articles.forEach((art) => {
    if (art.title) {
      const newObject = {
        author: art.author.username || 'Isaiah',
        profile: art.author.image || terrestial,
        title: art.title.length > 25 ? `${art.title.slice(0, 20)}...` : art.title,
        image: fetchImage(art.body),
        slug: art.slug || 'No-weher',
        id: art.id,
        date: `${new Date(art.createdAt).toDateString()}` || 'Fri Aug 30 2019',
      };
      newArticles.push(newObject);
    }
  });
  dispatch({
    type: GET_FEED,
    payload: newArticles,
    page: article.data.metadata.currentPage,
    totalPages: article.data.metadata.totalPages,
    nextPage: article.data.metadata.nextPage,
    previousPage: article.data.metadata.previousPage,
  });
  dispatch(setLoading(false));
};
