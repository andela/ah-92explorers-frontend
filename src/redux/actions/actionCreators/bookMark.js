import '@babel/polyfill';
import {
  toast,
} from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import axios from 'axios';
import {
  checkToken,
} from '../../../utils/checkToken';
import {
  BOOKMARK_ARTICLE_START,
  BOOKMARK_ARTICLE_SUCCESS,
  BOOKMARK_ARTICLE_FAILURE,
  BOOKMARK_GET_ALL,
  BOOKMARK_GET_ALL_FAIL,
  BOOKMARK_DELETE,
  BOOKMARK_DELETE_SUCCESS,
  BOOKMARK_DELETE_FAIL,
} from '../actionTypes/bookMark';

toast.configure();

const performAction = (type, payload) => ({
  type,
  payload,
});

export const bookmarkArticle = (slug) => async (dispatch) => {
  checkToken();
  dispatch(performAction(BOOKMARK_ARTICLE_START));
  try {
    const res = await axios.post(`${process.env.APP_URL_BACKEND}/api/articles/${slug}/bookmark`);
    dispatch(performAction(BOOKMARK_ARTICLE_SUCCESS, res.data));
  } catch (error) {
    dispatch(performAction(BOOKMARK_ARTICLE_FAILURE, error));
    if (Error) {
      toast.error('unauthorized to bookmark this article, please login/signup');
    }
  }
};


export const getAllBookmarks = () => async (dispatch) => {
  checkToken();
  try {
    const res = await axios.get(`${process.env.APP_URL_BACKEND}/api/bookmark`);
    return dispatch(performAction(BOOKMARK_GET_ALL, res.data));
  } catch (error) {
    return dispatch(performAction(BOOKMARK_GET_ALL_FAIL, error));
  }
};

export const removeBookmark = (slug) => async (dispatch) => {
  checkToken();
  try {
    const res = await axios.delete(`${process.env.APP_URL_BACKEND}/api/articles/${slug}/bookmark`);
    return dispatch(performAction(BOOKMARK_DELETE_SUCCESS, res.data));
  } catch (error) {
    return dispatch(performAction(BOOKMARK_DELETE_FAIL, error));
  }
};
