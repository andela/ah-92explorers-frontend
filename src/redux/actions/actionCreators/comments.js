/* eslint-disable import/no-cycle */
import dotenv from 'dotenv';
import axios from 'axios';
import { toast } from 'react-toastify';
import {
  SET_COMMENT_SUCCESS, SET_COMMENT_ERROR, GET_COMMENTS, SET_COMMENT_DELETE, EDIT_COMMENT_HISTORY,
} from '../actionTypes/comments';
import 'react-toastify/dist/ReactToastify.css';

dotenv.config();
toast.configure();

const networkError = 'Something went wrong. Check your internet connection.';

export function setCommentSuccess(comment) {
  return {
    type: SET_COMMENT_SUCCESS,
    payload: comment,
  };
}

export function setCommentError(commentError) {
  return {
    type: SET_COMMENT_ERROR,
    payload: commentError,
  };
}

export function getComments(comments) {
  return {
    type: GET_COMMENTS,
    payload: comments,
  };
}

export function setCommentDelete(isCommentDelete) {
  return {
    type: SET_COMMENT_DELETE,
    isCommentDelete,
  };
}

export function editCommentHistory(commentHistory) {
  return {
    type: EDIT_COMMENT_HISTORY,
    payload: commentHistory,
  };
}

export const postComment = (commentData, slug) => async (dispatch) => {
  try {
    const data = {
      body: commentData,
    };
    if (!data.body || data.body.length < 2) {
      return dispatch(setCommentError('Please add a valid comment'));
    }
    const res = await axios.post(`${process.env.APP_URL_BACKEND}/api/articles/${slug}/comments`, data, {
      headers: {
        Authorization: `Bearer ${localStorage.getItem('jwtToken')}`,
      },
    });
    if (res.data.message === 'commented') {
      dispatch(setCommentSuccess('You\'ve commented'));
      return data;
    }
  } catch (error) {
    let errorMessage = null;
    if (error.response) {
      errorMessage = error.response.data.error;
    }
    return dispatch(setCommentError(errorMessage));
  }
};

export const fetchComments = (slug) => async (dispatch) => {
  try {
    const comments = await axios.get(`${process.env.APP_URL_BACKEND}/api/articles/${slug}/comments`);
    return dispatch(getComments(comments.data.article.comments));
  } catch (error) {
    return dispatch(setCommentError(error));
  }
};

export const deleteComment = (commentId) => async (dispatch) => {
  try {
    await axios.delete(`${process.env.APP_URL_BACKEND}/api/comments/${commentId}`, {
      headers: {
        Authorization: `Bearer ${localStorage.getItem('jwtToken')}`,
      },
    });
    dispatch(setCommentDelete(true));
  } catch (error) {
    let errorMessage = null;
    if (error.response) {
      errorMessage = error.response.data.error;
    }
    dispatch(setCommentError(errorMessage));
    dispatch(setCommentDelete(false));
  }
};

export const updateComment = (commentData, id, lastComment) => async (dispatch) => {
  try {
    const data = {
      body: commentData,
    };
    if (!data.body || data.body.length < 2 || lastComment.trim() === commentData.trim()) {
      return dispatch(setCommentError('Please update to a new comment'));
    }
    const res = await axios.patch(`${process.env.APP_URL_BACKEND}/api/comments/${id}`, data, {
      headers: {
        Authorization: `Bearer ${localStorage.getItem('jwtToken')}`,
      },
    });
    if (res.data.message === 'successfully updated and tracked comment') {
      dispatch(setCommentSuccess('You\'ve updated the comment'));
      return data;
    }
  } catch (error) {
    let errorMessage = null;
    if (error.response) {
      errorMessage = error.response.data.error;
    } else {
      errorMessage = networkError;
    }
    return dispatch(setCommentError(errorMessage));
  }
};

export const fetchEditCommentHistory = (id) => async (dispatch) => {
  try {
    const editHistory = await axios.get(`${process.env.APP_URL_BACKEND}/api/comments/${id}`);
    return dispatch(editCommentHistory(editHistory.data.edits));
  } catch (error) {
    return dispatch(setCommentError(error.response.data.error));
  }
};

export const likeAComment = (id) => async (dispatch) => {
  try {
    const res = await axios.post(`${process.env.APP_URL_BACKEND}/api/article/comment/${id}/like`, {
      headers: {
        Authorization: `Bearer ${localStorage.getItem('jwtToken')}`,
      },
    });
    dispatch(setCommentSuccess(res.data.message));
  } catch (error) {
    let errorMessage = null;
    if (error.response) {
      errorMessage = error.response.data.error;
    }
    return dispatch(setCommentError(errorMessage));
  }
};
