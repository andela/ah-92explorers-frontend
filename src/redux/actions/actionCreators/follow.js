import '@babel/polyfill';
import { toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import axios from 'axios';
import dotenv from 'dotenv';

import {
  USER_FOLLOW, FOLLOW_SUCCESS, UNFOLLOW_SUCCESS, FOLLOW_FAILURE,
} from '../actionTypes';

dotenv.config();
toast.configure();
const { APP_URL_BACKEND } = process.env;

export const followOther = username => async (dispatch) => {
  try {
    dispatch({ type: USER_FOLLOW, payload: username });
    const data = await axios.post(`${APP_URL_BACKEND}/api/${username}/follow`,
      { headers: { Authorization: localStorage.getItem('jwtToken') } });
    if (data.status === 201) {
      dispatch({ type: FOLLOW_SUCCESS, payload: data.data.message });
      toast.success(data.data.message);
    }
    if (data.status === 200) {
      dispatch({ type: UNFOLLOW_SUCCESS, payload: data.data.message });
      toast.success(data.data.message);
    }
  } catch (error) {
    dispatch({ type: FOLLOW_FAILURE, payload: error });
    toast.error(error.response.data.error);
  }
};
