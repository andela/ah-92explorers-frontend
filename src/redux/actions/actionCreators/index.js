export * from './articles';
export * from './login';
export * from './signup';
export * from './home';
export * from './social.action';
