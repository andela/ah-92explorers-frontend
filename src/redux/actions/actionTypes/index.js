export * from './articles';
export * from './home';
export * from './login';
export * from './signup';
export * from './loading';
export * from './social.type';
