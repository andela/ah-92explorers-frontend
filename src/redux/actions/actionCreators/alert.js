import uuid from 'uuid';
import { SET_ALERT, REMOVE_ALERT } from '../actionTypes/alert';

export const setAlert = (msg, alertType, timeout = 5000) => (dispatch) => {
  const id = uuid.v4();
  dispatch({
    type: SET_ALERT,
    payload: { id, msg, alertType },
  });

  setTimeout(() => dispatch({ type: REMOVE_ALERT, payload: id }), timeout);
};
