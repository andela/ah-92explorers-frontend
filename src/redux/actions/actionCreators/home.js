import { GET_HOME } from '../actionTypes';

export function getHome() {
  return {
    type: GET_HOME,
    payload: 'Authors Haven in the Blime',
  };
}
