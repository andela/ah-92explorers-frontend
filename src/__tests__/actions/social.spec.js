import dotenv from 'dotenv';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { getInfo, success, failure } from '../../redux/actions/actionCreators';
import { socialTypes } from '../../redux/actions/actionTypes';

dotenv.config();
const middleWares = [thunk];
const mockStore = configureMockStore(middleWares);
let store = mockStore({});

const user = { firstName: 'Eleman', lastName: 'Hillary Barnabas', email: 'elemanhillary@gmail.com' };
describe('action creators', () => {
  it('should return LOGIN_SUCCESS action with user info', () => {
    const url = '?token=afeb658eb6b702e58baf3e1a41f7bb3e5911c1768ac241e83961605c76a4d2fc3bd53669057e5277621928e4b41cf1b72bf8eddf3a70e34babecfcb5e8ce0c2eeb63af45b32779cd176247e285fca02e5c8eb56331df3087091dd9c99fc5bad065dc0412b0827a799751b2e4a955f555abc06724f94b97fb10a7cc013fc81ed270bcb3c9ab3d3f37f279f10e0a94ea304bcbd42855e9244eea7bbdda7ac0acbe4387e1ee346f3e5ea4bc984dbcc409f789ee603a96ffded96b668794969425c24781a85adfe92d30570b2839ddfde0cb94f153a8606fa1d39dfc3f023be86b3975e76a6b6c2183c2013ebe342471b5cc6082c1af9a02baff6817e232f6b02cd819cc9ee0db008c13650e17d87308748ed777bc3743609439d3e797999e507394033b7dc1bb03e6dd520b48e56b3ad17b76c7e5ff1590e35bb0f66f3121de46cee5d2a39526740b002872c8b41fe3b48e272fe35a323ef25a5d7394de05853f453786912d4b1b35e73e5fe540ae9c862aec2a7512cc8a';
    delete window.location;
    window.location = { search: url };
    const expectedActions = [{
      type: socialTypes.LOGIN_SUCCESS,
      user,
    }];
    const searchURL = window.location.search;
    store.dispatch(getInfo(searchURL));
    expect(store.getActions()).toEqual(expectedActions);
  });
  it('should return LOGIN_FAILURE', () => {
    const url = '?token=bd53669057e5277621928e4b41cf1b72bf8eddf3a70e34babecfcb5e8ce0c2eeb63af45b32779cd176247e285fca02e5c8eb56331df3087091dd9c99fc5bad065dc0412b0827a799751b2e4a955f555abc06724f94b97fb10a7cc013fc81ed270bcb3c9ab3d3f37f279f10e0a94ea304bcbd42855e9244eea7bbdda7ac0acbe4387e1ee346f3e5ea4bc984dbcc409f789ee603a96ffded96b668794969425c24781a85adfe92d30570b2839ddfde0cb94f153a8606fa1d39dfc3f023be86b3975e76a6b6c2183c2013ebe342471b5cc6082c1af9a02baff6817e232f6b02cd819cc9ee0db008c13650e17d87308748ed777bc3743609439d3e797999e507394033b7dc1bb03e6dd520b48e56b3ad17b76c7e5ff1590e35bb0f66f3121de46cee5d2a39526740b002872c8b41fe3b48e272fe35a323ef25a5d7394de05853f453786912d4b1b35e73e5fe540ae9c862aec2a7512cc8a';
    delete window.location;
    window.location = { search: url };
    store = mockStore({});
    const expectedActions = [{
      type: socialTypes.LOGIN_FAILURE,
      error: 'something went wrong',
    }];
    const searchURL = window.location.search;
    store.dispatch(getInfo(searchURL));
    expect(store.getActions()).toEqual(expectedActions);
  });
  it('should return LOGIN_FAILURE with undefined token', () => {
    const url = '?token=';
    delete window.location;
    window.location = { search: url };
    const expectedActions = [{
      type: socialTypes.LOGIN_FAILURE,
      error: 'something went wrong',
    }];
    store = mockStore({});
    const searchURL = window.location.search;
    store.dispatch(getInfo(searchURL));
    expect(store.getActions()).toEqual(expectedActions);
  });
  it('creates LOGIN_SUCCESS action', () => {
    store = mockStore({});
    store.dispatch(success({ firstName: 'Eleman', lastName: 'Hillary Barnabas', email: 'elemanhillary@gmail.com' }));
    expect(store.getActions()).toMatchSnapshot();
  });
  it('creates LOGIN_FAILURE action', () => {
    store = mockStore({});
    store.dispatch(failure('something went wrong'));
    expect(store.getActions()).toMatchSnapshot();
  });
});
