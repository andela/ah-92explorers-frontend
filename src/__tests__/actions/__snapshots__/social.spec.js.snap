// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action creators creates LOGIN_FAILURE action 1`] = `
Array [
  Object {
    "error": "something went wrong",
    "type": "LOGIN_FAILURE",
  },
]
`;

exports[`action creators creates LOGIN_SUCCESS action 1`] = `
Array [
  Object {
    "type": "LOGIN_SUCCESS",
    "user": Object {
      "email": "elemanhillary@gmail.com",
      "firstName": "Eleman",
      "lastName": "Hillary Barnabas",
    },
  },
]
`;
